ResearchKit/Common/ORKPasscodeStepViewController.m

```diff
@@ -0,0 +1,571 @@
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+#import "ORKPasscodeStepViewController.h"
+#import "ORKPasscodeStepViewController_Internal.h"
+#import "ORKStepViewController_Internal.h"
+#import "ORKPasscodeStepView.h"
+#import "ORKPasscodeStep.h"
+#import "ORKKeychainWrapper.h"
+
+#import <AudioToolbox/AudioToolbox.h>
+#import <LocalAuthentication/LocalAuthentication.h>
+
+
+@implementation ORKPasscodeStepViewController {
+    ORKPasscodeStepView *_passcodeStepView;
+    NSMutableString *_passcode;
+    NSMutableString *_confirmPasscode;
+    NSInteger _position;
+    ORKPasscodeState _passcodeState;
+    BOOL _shouldResignFirstResponder;
+    BOOL _isChangingState;
+    BOOL _isTouchIdAuthenticated;
+    BOOL _isPasscodeSaved;
+    LAContext *_touchContext;
+}
+
+- (ORKPasscodeStep *)passcodeStep {
+    return (ORKPasscodeStep *)self.step;
+}
+
+- (void)stepDidChange {
+    [super stepDidChange];
+    
+    [_passcodeStepView removeFromSuperview];
+    _passcodeStepView = nil;
+    
+    if (self.step && [self isViewLoaded]) {
+        
+        _passcodeStepView = [[ORKPasscodeStepView alloc] initWithFrame:self.view.bounds];
+        _passcodeStepView.autoresizingMask = UIViewAutoresizingFlexibleWidth|UIViewAutoresizingFlexibleHeight;
+        _passcodeStepView.headerView.instructionLabel.text = [self passcodeStep].text;
+        _passcodeStepView.passcodeType = _passcodeType;
+        _passcodeStepView.textField.delegate = self;
+        [self.view addSubview:_passcodeStepView];
+        
+        _passcode = [NSMutableString new];
+        _confirmPasscode = [NSMutableString new];
+        _position = 0;
+        _shouldResignFirstResponder = NO;
+        _isChangingState = NO;
+        _isTouchIdAuthenticated = NO;
+        _isPasscodeSaved = NO;
+        
+        // Set the starting passcode state based on flow.
+        switch (_passcodeFlow) {
+            case ORKPasscodeFlowCreate:
+                _passcodeState = ORKPasscodeStateEntry;
+                [self updatePasscodeView];
+                break;
+            
+            case ORKPasscodeFlowAuthenticate:
+                _passcodeState = ORKPasscodeStateEntry;
+                [self updatePasscodeView];
+                break;
+                
+            case ORKPasscodeFlowEdit:
+                _passcodeState = ORKPasscodeStateOldEntry;
+                [self updatePasscodeView];
+                break;
+        }
+        
+        // If creating a new passcode, clear out the keychain.
+        if (self.passcodeFlow == ORKPasscodeFlowCreate) {
+            [self removePasscodeFromKeychain];
+            _useTouchId = YES;
```

Comment:
If we force dev to use TouchID, would it be more safe to override `useTouchId` getter to return YES when passcodeFlow is `ORKPasscodeFlowCreate`? So it won't be overridden by accident.